esphome/components/tx20/tx20.cpp

```diff
@@ -0,0 +1,195 @@
+    // ignore segments at the end that were too short
+    string_buffer.append(repeat, current_bit ? '1' : '0');
+    bit_buffer.insert(bit_buffer.end(), repeat, current_bit);
+    current_bit = !current_bit;
+  }
+  current_bit = !current_bit;
+  if (string_buffer.length() < MAX_BUFFER_SIZE) {
+    uint8_t remain = MAX_BUFFER_SIZE - string_buffer.length();
+    string_buffer_2 += to_string(remain) + ", ";
+    string_buffer.append(remain, current_bit ? '1' : '0');
+    bit_buffer.insert(bit_buffer.end(), remain, current_bit);
+  }
+
+  uint8_t tx20_sa = 0;
+  uint8_t tx20_sb = 0;
+  uint8_t tx20_sd = 0;
+  uint8_t tx20_se = 0;
+  uint16_t tx20_sc = 0;
+  uint16_t tx20_sf = 0;
+  uint8_t tx20_wind_direction = 0;
+  float tx20_wind_speed_kmh = 0;
+  uint8_t bit_count = 0;
+
+  for (int i = 41; i > 0; i--) {
+    uint8_t bit = bit_buffer.at(bit_count);
+    bit_count++;
+    if (i > 41 - 5) {
+      // start, inverted
+      tx20_sa = (tx20_sa << 1) | (bit ^ 1);
+    } else if (i > 41 - 5 - 4) {
+      // wind dir, inverted
+      tx20_sb = tx20_sb >> 1 | ((bit ^ 1) << 3);
+    } else if (i > 41 - 5 - 4 - 12) {
+      // windspeed, inverted
+      tx20_sc = tx20_sc >> 1 | ((bit ^ 1) << 11);
+    } else if (i > 41 - 5 - 4 - 12 - 4) {
+      // checksum, inverted
+      tx20_sd = tx20_sd >> 1 | ((bit ^ 1) << 3);
+    } else if (i > 41 - 5 - 4 - 12 - 4 - 4) {
+      // wind dir
+      tx20_se = tx20_se >> 1 | (bit << 3);
+    } else {
+      // windspeed
+      tx20_sf = tx20_sf >> 1 | (bit << 11);
+    }
+  }
+
+  uint8_t chk = (tx20_sb + (tx20_sc & 0xf) + ((tx20_sc >> 4) & 0xf) + ((tx20_sc >> 8) & 0xf));
+  chk &= 0xf;
+  bool value_set = false;
+  // checks:
+  // 1. Check that the start frame is 00100 (0x04)
+  // 2. Check received checksum matches calculated checksum
+  // 3. Check that Wind Direction matches Wind Direction (Inverted)
+  // 4. Check that Wind Speed matches Wind Speed (Inverted)
+  ESP_LOGVV(TAG, "BUFFER %s", string_buffer_2.c_str());
+  ESP_LOGVV(TAG, "Decoded bits %s", string_buffer.c_str());
+
+  if (tx20_sa == 4) {
+    if (chk == tx20_sd) {
+      if (tx20_sf == tx20_sc) {
+        tx20_wind_speed_kmh = float(tx20_sc) * 0.36;
+        ESP_LOGV(TAG, "WindSpeed %f", tx20_wind_speed_kmh);
+        if (this->wind_speed_sensor_ != nullptr)
+          this->wind_speed_sensor_->publish_state(tx20_wind_speed_kmh);
+        value_set = true;
+      }
+      if (tx20_se == tx20_sb) {
+        tx20_wind_direction = tx20_se;
+        if (tx20_wind_direction >= 0 && tx20_wind_direction < 16) {
+          wind_cardinal_direction_ = DIRECTIONS[tx20_wind_direction];
+        }
+        ESP_LOGV(TAG, "WindDirection %d", tx20_wind_direction);
+        if (this->wind_direction_degrees_sensor_ != nullptr)
+          this->wind_direction_degrees_sensor_->publish_state(float(tx20_wind_direction) * 22.5f);
+        value_set = true;
+      }
+      if (!value_set) {
+        ESP_LOGE(TAG, "No value set!");
```

Comment:
```suggestion
        ESP_LOGW(TAG, "No value set!");
```

Error is for unrecoverable errors - these should all be warnings as the rest of the device can continue to function

Reply:
Makes sense - fixed!